lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,378 @@
+            "GroundTypeAnnotation was found on a non-ground type.");
+      }
+
+      auto component = &(*interfaceMap)[{defName, name}];
+      component->found |= true;
+
+      switch (component->tpe) {
+      case Component::Type::Bundle:
+        // Do nothing.
+        break;
+      case Component::Type::Vector:
+        component->width = type.getBitWidthOrSentinel();
+        component->depth++;
+        break;
+      case Component::Type::Ground:
+        component->width = type.getBitWidthOrSentinel();
+        break;
+      case Component::Type::Unknown:
+        llvm_unreachable("Shouldn't be here");
+      }
+
+      continue;
+    }
+
+    unprocessedAnnos.push_back(a);
+  }
+
+  if (unprocessedAnnos.empty())
+    return {};
+  else
+    return ArrayAttr::get(context, unprocessedAnnos);
+};
+
+/// Combined logic to handle both Wires, Registers, and Nodes because these all
+/// use the same approach.
+void GrandCentralVisitor::handleRef(mlir::Operation *op) {
+
+  auto remainingAnnotations = handleRefLike(
+      op, getAnnotations(op), op->getResult(0).getType().cast<FIRRTLType>());
+
+  if (!remainingAnnotations || remainingAnnotations.empty())
+    op->removeAttr("annotations");
+  else
+    op->setAttr("annotations", remainingAnnotations);
+};
+
+/// Remove Grand Central Annotations from ports of modules or external modules.
+/// Return argument attributes with annotations removed.
+ArrayAttr GrandCentralVisitor::handlePorts(Operation *op) {
+
+  SmallVector<Attribute> newArgAttrs;
+  auto ports = getModulePortInfo(op);
+  for (size_t i = 0, e = ports.size(); i != e; ++i) {
+    auto port = ports[i];
+    auto argAttr = mlir::function_like_impl::getArgAttrs(op, i);
+    if (!port.annotations) {
+      newArgAttrs.push_back(DictionaryAttr::get(op->getContext(), argAttr));
+      continue;
+    }
+
+    auto remainingAnnotations = handleRefLike(op, port.annotations, port.type);
+
+    // Erase the annotations that we consumed.
+    SmallVector<NamedAttribute> newArgAttr;
+    for (auto argIter = argAttr.begin(), argEnd = argAttr.end();
+         argIter != argEnd; ++argIter) {
+      if (argIter->first != "firrtl.annotations") {
+        newArgAttr.push_back(*argIter);
+        continue;
+      }
+      if (!remainingAnnotations || remainingAnnotations.empty())
+        continue;
+      newArgAttr.push_back({argIter->first, remainingAnnotations});
+    }
+    newArgAttrs.push_back(DictionaryAttr::get(op->getContext(), newArgAttr));
+  }
+
+  return ArrayAttr::get(op->getContext(), newArgAttrs);
+};
+
+void GrandCentralPass::runOnOperation() {
+  CircuitOp circuitOp = getOperation();
+
+  ArrayAttr annotations = circuitOp->getAttrOfType<ArrayAttr>("annotations");
+  if (!annotations)
+    return;
+
+  auto context = &getContext();
+  OpBuilder builder(context);
+  builder.setInsertionPointToEnd(getOperation()->getBlock());
+
+  Options opts;
+
+  opts.extractGCTAnno = StringAttr::get(
```

Comment:
"Options" isn't really the right name for this struct, but bigger comment:

I don't think that comparing uniqued StringAttr's are worth it.  It increases complexity of the code in the compiler, and the microoptimization (pointer comparison of attrs, instead of strcmp of identifiers) is dubious at best.  These are sparse attributes and the strcmp of the identifier starts with a length check.  I'd just use the `Annotation::isClass` member and be done with it.

Such an approach defines away the Options struct and simplifies the code.

Reply:
Changed to just use strings internally.